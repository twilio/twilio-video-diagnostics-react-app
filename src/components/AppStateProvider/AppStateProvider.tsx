import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
