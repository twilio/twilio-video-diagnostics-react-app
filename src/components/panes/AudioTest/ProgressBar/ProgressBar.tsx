import { useRef, useEffect } from 'react';
import { alpha } from '@material-ui/core/styles/colorManipulator';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
  container: {
    position: 'relative',
    margin: '0 0.2em 0',
    height: '4px',
    '& div': {
      position: 'absolute',
      top: 0,
      bottom: 0,
      left: 0,
    },
  },
  progress: {
    background: '#14B053',
    right: '100%',
  },
  background: {
    right: 0,
    background: alpha('#E1E3EA', 0.2),
  },
});

interface ProgressBarProps {
  duration: number;
  position: number;
  style?: { [key: string]: string };
}

export default function ProgressBar({ duration, position, style }: ProgressBarProps) {
  const classes = useStyles();
  const progressBarRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const el = progressBarRef.current;
    if (el) {
      window.requestAnimationFrame(() => {
        // We set these values asynchronously so that the browser can recognize the change in the 'right' value.
        // Without this, the progress bar would instantly snap to the designated position.
        el.style.transition = `right ${duration}s linear`;
        el.style.right = `${Math.max(0, 100 - position)}%`;
      });
    }
  }, [duration, position]);

  return (
    <div className={classes.container} style={{ ...style }} data-testid="progressBarContainer">
      <div className={classes.progress} ref={progressBarRef} data-testid="progressBar"></div>
      <div className={classes.background}></div>
    </div>
  );
}
