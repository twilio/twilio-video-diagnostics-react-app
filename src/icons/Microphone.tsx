import React from 'react';

function Microphone() {
  return (
    <svg width="12" height="17" viewBox="0 0 12 17" fill="none" xmlns="http://www.w3.org/2000/svg">
      <path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M3.11111 2.83333C3.11111 1.26812 4.37982 0 5.94444 0C7.51014 0 8.77778 1.26764 8.77778 2.83333V7.16667C8.77778 8.73129 7.50966 10 5.94444 10C4.3803 10 3.11111 8.73081 3.11111 7.16667V2.83333ZM7.77778 2.83333C7.77778 1.81992 6.95786 1 5.94444 1C4.932 1 4.11111 1.82051 4.11111 2.83333V7.16667C4.11111 8.17852 4.93259 9 5.94444 9C6.95727 9 7.77778 8.17911 7.77778 7.16667V2.83333ZM11.3889 6.66667C11.665 6.66667 11.8889 6.89052 11.8889 7.16667C11.8889 10.2813 9.49349 12.8365 6.44444 13.0904V15.7222C6.44444 15.9984 6.22059 16.2222 5.94444 16.2222C5.69898 16.2222 5.49484 16.0453 5.4525 15.8121L5.44444 15.7222V13.0904C2.3954 12.8365 0 10.2813 0 7.16667C0 6.89052 0.223858 6.66667 0.5 6.66667C0.776142 6.66667 1 6.89052 1 7.16667C1 9.89741 3.2137 12.1111 5.94444 12.1111C8.67519 12.1111 10.8889 9.89741 10.8889 7.16667C10.8889 6.89052 11.1127 6.66667 11.3889 6.66667Z"
        fill="#4B5671"
      />
    </svg>
  );
}

export default React.memo(Microphone);
