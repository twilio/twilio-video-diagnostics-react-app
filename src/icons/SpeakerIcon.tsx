import React from 'react';

function SpeakerIcon() {
  return (
    <svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
      <path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M9.12991 4.22311L5.26525 7.71535H3.41149C2.62537 7.71535 2 8.24395 2 9.00737V10.8531L2.00578 10.9807C2.0708 11.6993 2.67643 12.3297 3.41149 12.3297H5.26619L9.14319 15.7851C9.355 15.9926 9.69858 16.0603 9.99016 15.9419C10.2851 15.8223 10.4683 15.5399 10.4689 15.2313L10.4689 4.77139C10.4683 4.46179 10.2851 4.17862 9.99016 4.05894C9.69858 3.94059 9.355 4.00305 9.12991 4.22311ZM9.52795 5.11655V14.9267L5.76921 11.529C5.68232 11.4504 5.56848 11.4068 5.45031 11.4068H3.41149C3.18194 11.4068 2.94099 11.1379 2.94099 10.8531V9.00737C2.94099 8.78663 3.11657 8.63822 3.41149 8.63822H5.45031C5.56848 8.63822 5.68232 8.59461 5.76921 8.51606L9.52795 5.11655ZM15.6532 4.66938C15.4624 4.49642 15.1647 4.50793 14.9884 4.69509C14.812 4.88225 14.8237 5.17419 15.0146 5.34715C16.3179 6.5284 17.059 8.18978 17.059 9.93024C17.059 11.6707 16.3179 13.3321 15.0146 14.5133C14.8237 14.6863 14.812 14.9782 14.9884 15.1654C15.1647 15.3526 15.4624 15.3641 15.6532 15.1911C17.1493 13.8352 18 11.9281 18 9.93024C18 7.93239 17.1493 6.02532 15.6532 4.66938ZM13.0961 5.95963C13.2533 5.7567 13.5484 5.71716 13.7554 5.87132C15.049 6.83518 15.8087 8.33688 15.8087 9.93024C15.8087 11.5236 15.049 13.0253 13.7554 13.9892C13.5484 14.1433 13.2533 14.1038 13.0961 13.9009C12.9389 13.6979 12.9792 13.4085 13.1861 13.2543C14.2456 12.4649 14.8678 11.2351 14.8678 9.93024C14.8678 8.62536 14.2456 7.39554 13.1861 6.60619C12.9792 6.45203 12.9389 6.16255 13.0961 5.95963ZM12.1572 7.4037C11.947 7.25381 11.6528 7.29937 11.4999 7.50546C11.3471 7.71156 11.3936 8.00015 11.6037 8.15005C13.0228 9.1623 13.0216 10.6997 11.6038 11.7104C11.3936 11.8602 11.3471 12.1488 11.4999 12.3549C11.6526 12.5611 11.9469 12.6067 12.1571 12.4569C14.0915 11.0779 14.0933 8.78475 12.1572 7.4037Z"
        fill="#606B85"
      />
    </svg>
  );
}

export default React.memo(SpeakerIcon);
