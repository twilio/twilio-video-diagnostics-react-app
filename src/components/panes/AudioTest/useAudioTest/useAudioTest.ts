import { useState, useCallback } from 'react';
import { getLogger } from 'loglevel';
import {
  AudioInputTest,
  AudioOutputTest,
  DiagnosticError,
  testAudioInputDevice,
  testAudioOutputDevice,
  WarningName,
} from '@twilio/rtc-diagnostics';

import { getAudioLevelPercentage, getStandardDeviation } from '../../../../utils';
import { APP_NAME, INPUT_TEST_DURATION, RECORD_DURATION } from '../../../../constants';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';

const log = getLogger(APP_NAME);
let audioInputTest: AudioInputTest;
let audioOutputTest: AudioOutputTest;

const getErrorMessage = (error: DiagnosticError) => {
  let message = 'An unknown error has occurred';
  if (error) {
    message = error.domError ? error.domError.toString() : error.message;
  }
  return message;
};

export default function useAudioTest() {
  const [isRecording, setIsRecording] = useState(false);
  const [isAudioInputTestRunning, setIsAudioInputTestRunning] = useState(false);
  const [isAudioOutputTestRunning, setIsAudioOutputTestRunning] = useState(false);
  const [inputLevel, setInputLevel] = useState(0);
  const [outputLevel, setOutputLevel] = useState(0);
  const [playbackURI, setPlaybackURI] = useState('');
  const [error, setError] = useState('');
  const { dispatch } = useAppStateContext();

  const playAudio = useCallback(
    (options: AudioOutputTest.Options) => {
      log.debug('AudioOutputTest running');

      options = { doLoop: false, ...options };
      audioOutputTest = testAudioOutputDevice(options);
      setIsAudioOutputTestRunning(true);

      audioOutputTest.on(AudioOutputTest.Events.Volume, (value: number) => {
        setOutputLevel(getAudioLevelPercentage(value));
      });

      audioOutputTest.on(AudioOutputTest.Events.End, (report: AudioOutputTest.Report) => {
        setIsAudioOutputTestRunning(false);
        setOutputLevel(0);

        const stdDev = getStandardDeviation(report.values);
        if (stdDev === 0) {
          setError('No audio detected');
        }
        log.debug('AudioOutputTest ended', report);
        dispatch({ type: 'set-audio-output-test-report', report: report });
      });

      audioOutputTest.on(AudioOutputTest.Events.Error, (diagnosticError: DiagnosticError) => {
        log.debug('error', diagnosticError);
        setError(getErrorMessage(diagnosticError));
      });
    },
    [dispatch]
  );

  const readAudioInput = useCallback(
    (options: AudioInputTest.Options) => {
      if (audioInputTest) {
        audioInputTest.stop();
      }

      log.debug('AudioInputTest running');
      const duration = options.enableRecording ? RECORD_DURATION : INPUT_TEST_DURATION;
      options = { duration, ...options };
      audioInputTest = testAudioInputDevice(options);

      setIsAudioInputTestRunning(true);
      if (options.enableRecording) {
        log.debug('Recording audio');
        setIsRecording(true);
      }

      audioInputTest.on(AudioInputTest.Events.Volume, (value: number) => {
        setInputLevel(getAudioLevelPercentage(value));
      });

      audioInputTest.on(AudioInputTest.Events.End, (report: AudioInputTest.Report) => {
        if (playbackURI && report.recordingUrl) {
          URL.revokeObjectURL(playbackURI);
        }

        if (report.recordingUrl) {
          setPlaybackURI(report.recordingUrl);
        }

        setIsRecording(false);
        setIsAudioInputTestRunning(false);
        log.debug('AudioInputTest ended', report);
        dispatch({ type: 'set-audio-input-test-report', report: report });
      });

      audioInputTest.on(AudioInputTest.Events.Error, (diagnosticError: DiagnosticError) => {
        log.debug('error', diagnosticError);
        setError(getErrorMessage(diagnosticError));
      });
      audioInputTest.on(AudioInputTest.Events.Warning, (name: WarningName) => {
        log.debug('warning', name);
      });
      audioInputTest.on(AudioInputTest.Events.WarningCleared, (name: WarningName) => {
        log.debug('warning-cleared', name);
      });
    },
    [playbackURI, dispatch]
  );

  const stopAudioTest = useCallback(() => {
    if (audioInputTest) {
      audioInputTest.stop();
      setInputLevel(0);
    }
    if (audioOutputTest) {
      audioOutputTest.stop();
      setOutputLevel(0);
    }
  }, []);

  return {
    error,
    isRecording,
    isAudioInputTestRunning,
    isAudioOutputTestRunning,
    playAudio,
    playbackURI,
    readAudioInput,
    stopAudioTest,
    inputLevel,
    outputLevel,
  };
}
