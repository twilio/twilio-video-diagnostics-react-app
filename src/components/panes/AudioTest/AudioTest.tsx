import { useEffect, useRef, useState } from 'react';
import { Button, Paper, Typography, Container, Grid, Theme, createStyles } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import AudioDevice from './AudioDevice/AudioDevice';
import ProgressBar from './ProgressBar/ProgressBar';
import useAudioTest from './useAudioTest/useAudioTest';
import { ActivePane, useAppStateContext } from '../../AppStateProvider/AppStateProvider';
import Microphone from '../../../icons/Microphone';
import Speaker from '../../../icons/SpeakerIcon';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    mainContainer: {
      display: 'block',
    },
    paper: {
      display: 'inline-block',
      padding: '23px',
      borderRadius: '8px',
      minHeight: '280px',
      maxWidth: '365px',
      [theme.breakpoints.down('md')]: {
        width: '100%',
      },
    },
    audioLevelContainer: {
      display: 'flex',
      alignItems: 'center',
      height: '2.5em',
    },
    topLine: {
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'baseline',
    },
    header: {
      float: 'left',
      [theme.breakpoints.down('md')]: {
        float: 'initial',
      },
    },
    audioTest: {
      float: 'right',
      marginRight: '1em',
      [theme.breakpoints.down('md')]: {
        float: 'initial',
        display: 'flex',
        justifyContent: 'center',
        margin: '0 0 2em 0',
      },
    },
    confirmationButtons: {
      clear: 'left',
      [theme.breakpoints.down('md')]: {
        clear: 'initial',
        marginBottom: '2em',
      },
    },
  })
);

export function AudioTest() {
  const classes = useStyles();
  const [inputDeviceId, setInputDeviceId] = useState('');
  const [outputDeviceId, setOutputDeviceId] = useState('');
  const previousInputDeviceIdRef = useRef('');
  const { state, dispatch } = useAppStateContext();

  const {
    error,
    setError,
    isRecording,
    isAudioInputTestRunning,
    isAudioOutputTestRunning,
    playAudio,
    playbackURI,
    readAudioInput,
    stopAudioTest,
    inputLevel,
    outputLevel,
  } = useAudioTest();

  const volumeLevel = isAudioOutputTestRunning ? outputLevel : inputLevel;

  const disableAll = isRecording || isAudioOutputTestRunning || (!!error && error !== 'No audio detected');

  const handleRecordClick = () => {
    readAudioInput({ deviceId: inputDeviceId, enableRecording: true });
  };

  const handlePlayClick = () => {
    playAudio({ deviceId: outputDeviceId, testURI: playbackURI });
  };

  // stop test when not on AudioTest and there's an active test
  useEffect(() => {
    if (state.activePane !== ActivePane.AudioTest && (isAudioOutputTestRunning || isAudioInputTestRunning)) {
      stopAudioTest();
    }
  }, [state.activePane, stopAudioTest, isAudioInputTestRunning, isAudioOutputTestRunning]);

  // start audio test when on AudioTest and deviceId changes
  useEffect(() => {
    if (state.activePane === ActivePane.AudioTest) {
      const newInputDeviceSelected = previousInputDeviceIdRef.current !== inputDeviceId;
      previousInputDeviceIdRef.current = inputDeviceId;

      // Restarts the test to continuously capture audio input
      if (!error && (newInputDeviceSelected || (!isRecording && !isAudioInputTestRunning))) {
        readAudioInput({ deviceId: inputDeviceId });
      }
    }
    if (error) {
      stopAudioTest();
    }
  }, [
    error,
    state.activePane,
    inputDeviceId,
    isRecording,
    isAudioInputTestRunning,
    readAudioInput,
    dispatch,
    stopAudioTest,
  ]);

  return (
    <Container>
      <div className={classes.mainContainer}>
        <Grid item lg={5} className={classes.header}>
          <Typography variant="h1" gutterBottom>
            Test your audio
          </Typography>

          <Typography variant="body1" gutterBottom>
            Record an audio clip and play it back to check that your microphone and speaker are working. If they aren't,
            make sure your volume is turned up, try a different speaker or microphone, or check your bluetooth settings.
          </Typography>
        </Grid>

        <Grid item lg={5} className={classes.audioTest}>
          <Paper className={classes.paper}>
            <div className={classes.topLine}>
              <Typography variant="subtitle2">
                <strong>Audio</strong>
              </Typography>
              <div>
                <Button
                  variant="outlined"
                  style={{ marginRight: '1em' }}
                  size="small"
                  disabled={disableAll}
                  onClick={handleRecordClick}
                >
                  Record
                </Button>
                <Button variant="outlined" size="small" disabled={!playbackURI || disableAll} onClick={handlePlayClick}>
                  Play back
                </Button>
              </div>
            </div>
            â€‹
            <AudioDevice
              disabled={disableAll}
              kind="audiooutput"
              onDeviceChange={setOutputDeviceId}
              setDeviceError={setError}
            />
            <AudioDevice
              disabled={disableAll}
              kind="audioinput"
              onDeviceChange={setInputDeviceId}
              setDeviceError={setError}
              error={error}
            />
            <div className={classes.audioLevelContainer}>
              <div style={{ width: '2em', display: 'flex', justifyContent: 'center' }}>
                {isAudioOutputTestRunning ? <Speaker /> : <Microphone />}
              </div>
              <ProgressBar position={volumeLevel} duration={0.1} style={{ flex: '1' }} />
            </div>
          </Paper>
        </Grid>

        <Grid item lg={5} className={classes.confirmationButtons}>
          <Typography variant="body1" gutterBottom>
            <strong> Does your audio sound good?</strong>
          </Typography>

          <Button
            variant="contained"
            style={{ marginRight: '1.5em' }}
            color="primary"
            onClick={() => dispatch({ type: 'next-pane' })}
            disabled={!!error && error !== 'No audio detected'}
          >
            Yes
          </Button>

          <Button
            color="primary"
            onClick={() => dispatch({ type: 'next-pane' })}
            disabled={!!error && error !== 'No audio detected'}
          >
            Skip for now
          </Button>
        </Grid>
      </div>
    </Container>
  );
}
