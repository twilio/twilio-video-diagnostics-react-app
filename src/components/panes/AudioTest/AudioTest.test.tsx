import { Button } from '@material-ui/core';
import { mount, shallow } from 'enzyme';

import { AudioTest } from './AudioTest';
import { ActivePane, useAppStateContext } from '../../AppStateProvider/AppStateProvider';
import Microphone from '../../../icons/Microphone';
import SpeakerIcon from '../../../icons/SpeakerIcon';
import ProgressBar from './ProgressBar/ProgressBar';
import useAudioTest from './useAudioTest/useAudioTest';
import useDevices from '../../../hooks/useDevices/useDevices';

jest.mock('../../AppStateProvider/AppStateProvider');
jest.mock('./useAudioTest/useAudioTest');
jest.mock('../../../hooks/useDevices/useDevices');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;
const mockUseAudioTest = useAudioTest as jest.Mock<any>;
const mockUseDevices = useDevices as jest.Mock<any>;

const mockDispatch = jest.fn();

mockUseAppStateContext.mockImplementation(() => ({
  state: { activePane: ActivePane.AudioTest },
  dispatch: mockDispatch,
}));

mockUseDevices.mockImplementation(() => ({
  audioInputDevices: [
    { deviceId: 1, label: 'mockInput1' },
    { deviceId: 2, label: 'mockInput2' },
  ],
  videoInputDevices: [],
  audioOutputDevices: [
    { deviceId: 3, label: 'mockOutput3' },
    { deviceId: 4, label: 'mockOutput4' },
  ],
}));

describe('the AudioTest component', () => {
  let hookProps: any;

  beforeEach(() => {
    hookProps = {
      error: '',
      inputLevel: 0,
      isRecording: false,
      isAudioInputTestRunning: false,
      isAudioOutputTestRunning: false,
      outputLevel: 0,
      playAudio: jest.fn(),
      playbackURI: '',
      readAudioInput: jest.fn(),
      testEnded: false,
      stopAudioTest: jest.fn(),
      setError: jest.fn(),
    };
    mockUseAudioTest.mockImplementation(() => hookProps);
  });

  it('should render correctly', () => {
    const wrapper = shallow(<AudioTest />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should stop the test when active pane is not AudioTest and there is a test in progress', () => {
    mockUseAudioTest.mockImplementationOnce(() => ({ ...hookProps, isAudioOutputTestRunning: true }));
    mockUseAppStateContext.mockImplementationOnce(() => ({ state: { activePane: ActivePane.Connectivity } }));

    mount(<AudioTest />);

    expect(hookProps.stopAudioTest).toHaveBeenCalled();
  });

  it('should stop the test when there is an audio test error', () => {
    mockUseAudioTest.mockImplementationOnce(() => ({
      ...hookProps,
      isAudioOutputTestRunning: true,
      error: 'mockError',
    }));

    mount(<AudioTest />);

    expect(hookProps.stopAudioTest).toHaveBeenCalled();
  });

  describe('passive testing', () => {
    it('should start passive testing by default', () => {
      mount(<AudioTest />);
      expect(hookProps.readAudioInput).toHaveBeenCalledWith({ deviceId: '' });
    });

    [
      {
        shouldBeCalled: true,
        props: { error: '', isRecording: false, isAudioInputTestRunning: false },
      },
      {
        shouldBeCalled: false,
        props: { error: '', isRecording: true, isAudioInputTestRunning: false },
      },
      {
        shouldBeCalled: false,
        props: { error: '', isRecording: false, isAudioInputTestRunning: true },
      },
      {
        shouldBeCalled: false,
        props: { error: '', isRecording: true, isAudioInputTestRunning: true },
      },
      {
        shouldBeCalled: false,
        props: { error: 'foo', isRecording: false, isAudioInputTestRunning: false },
      },
      {
        shouldBeCalled: false,
        props: { error: 'foo', isRecording: true, isAudioInputTestRunning: false },
      },
      {
        shouldBeCalled: false,
        props: { error: 'foo', isRecording: false, isAudioInputTestRunning: true },
      },
      {
        shouldBeCalled: false,
        props: { error: 'foo', isRecording: true, isAudioInputTestRunning: true },
      },
    ].forEach(({ shouldBeCalled, props }) => {
      it(`should${shouldBeCalled ? ' ' : ' not '}call readAudioInput when props are ${JSON.stringify(props)}`, () => {
        hookProps = { ...hookProps, ...props };
        mount(<AudioTest />);

        if (shouldBeCalled) {
          expect(hookProps.readAudioInput).toHaveBeenCalledWith({ deviceId: '' }); // eslint-disable-line
        } else {
          expect(hookProps.readAudioInput).not.toHaveBeenCalledWith({ deviceId: '' }); // eslint-disable-line
        }
      });
    });
  });

  describe('button clicks', () => {
    beforeEach(() => {
      hookProps = { ...hookProps, isAudioInputTestRunning: true, playbackURI: 'foo' };
    });

    it('should start recording when "Record" button is clicked', () => {
      const wrapper = mount(<AudioTest />);
      const recordBtn = wrapper.find(Button).at(0);
      recordBtn.simulate('click');
      expect(hookProps.readAudioInput).toHaveBeenCalledWith({ deviceId: 1, enableRecording: true });
    });

    it('should play recorded message when "Play" button is clicked', () => {
      const wrapper = mount(<AudioTest />);
      const playBtn = wrapper.find(Button).at(1);
      playBtn.simulate('click');
      expect(hookProps.playAudio).toHaveBeenCalledWith({ deviceId: 3, testURI: 'foo' });
    });

    it('should go the next pane when "Yes" button is clicked', () => {
      const wrapper = mount(<AudioTest />);
      const yesBtn = wrapper.find(Button).at(2);
      yesBtn.simulate('click');
      expect(mockDispatch).toHaveBeenCalledWith({ type: 'next-pane' });
    });

    it('should go the next pane when "Skip for now" button is clicked', () => {
      const wrapper = mount(<AudioTest />);
      const skipBtn = wrapper.find(Button).at(3);
      skipBtn.simulate('click');
      expect(mockDispatch).toHaveBeenCalledWith({ type: 'next-pane' });
    });

    it('should disable the "Yes" and "Skip for now" buttons when there is an error that is not "No audio detected"', () => {
      mockUseAudioTest.mockImplementation(() => ({ ...hookProps, isAudioInputTestRunning: true, error: 'mockError' }));

      const wrapper = mount(<AudioTest />);
      const yesBtn = wrapper.find(Button).at(2);
      const skipBtn = wrapper.find(Button).at(3);

      expect(yesBtn.prop('disabled')).toBe(true);
      expect(skipBtn.prop('disabled')).toBe(true);
    });

    it('should not disable buttons when there is an error that is not "No audio detected"', () => {
      mockUseAudioTest.mockImplementation(() => ({
        ...hookProps,
        isAudioInputTestRunning: true,
        error: 'No audio detected',
      }));

      const wrapper = mount(<AudioTest />);
      const yesBtn = wrapper.find(Button).at(0);
      const skipBtn = wrapper.find(Button).at(1);

      expect(yesBtn.prop('disabled')).toBe(false);
      expect(skipBtn.prop('disabled')).toBe(false);
    });
  });

  describe('button labels', () => {
    it('should set record button label to "Record"', () => {
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(Button).at(0).text()).toEqual('Record');
    });

    it('should set play button label to "Play back"', () => {
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(Button).at(1).text()).toEqual('Play back');
    });
  });

  describe('volume levels', () => {
    it('should pass inputLevel to ProgressBar when isAudioOutputTestRunning is false', () => {
      hookProps = { ...hookProps, inputLevel: 64, outputLevel: 0, isAudioOutputTestRunning: false };
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(ProgressBar).props().position).toEqual(64);
    });

    it('should pass outputLevel to ProgressBar when isAudioOutputTestRunning is true', () => {
      hookProps = { ...hookProps, inputLevel: 64, outputLevel: 93, isAudioOutputTestRunning: true };
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(ProgressBar).props().position).toEqual(93);
    });

    it('should display the microphone icon when isAudioOutputTestRunning is false', () => {
      hookProps = { ...hookProps, inputLevel: 64, outputLevel: 0, isAudioOutputTestRunning: false };
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(Microphone).exists()).toBe(true);
    });

    it('should display the speaker icon when isAudioOutputTestRunning is true', () => {
      hookProps = { ...hookProps, inputLevel: 64, outputLevel: 93, isAudioOutputTestRunning: true };
      const wrapper = shallow(<AudioTest />);
      expect(wrapper.find(SpeakerIcon).exists()).toBe(true);
    });
  });
});
