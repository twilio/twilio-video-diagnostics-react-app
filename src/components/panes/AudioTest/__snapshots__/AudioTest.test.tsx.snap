// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the AudioTest component should render correctly 1`] = `
<WithStyles(ForwardRef(Container))>
  <div
    className="makeStyles-mainContainer-1"
  >
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-header-5"
      item={true}
      lg={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h1"
      >
        Test your audio
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        Record an audio clip and play it back to check that your microphone and speaker are working. If they aren't, make sure your volume is turned up, try a different speaker or microphone, or check your bluetooth settings.
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-audioTest-6"
      item={true}
      lg={5}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <div
          className="makeStyles-topLine-4"
        >
          <WithStyles(ForwardRef(Typography))
            variant="subtitle2"
          >
            <strong>
              Audio
            </strong>
          </WithStyles(ForwardRef(Typography))>
          <div>
            <WithStyles(ForwardRef(Button))
              disabled={false}
              onClick={[Function]}
              size="small"
              style={
                Object {
                  "marginRight": "1em",
                }
              }
              variant="outlined"
            >
              Record
            </WithStyles(ForwardRef(Button))>
            <WithStyles(ForwardRef(Button))
              disabled={true}
              onClick={[Function]}
              size="small"
              variant="outlined"
            >
              Play back
            </WithStyles(ForwardRef(Button))>
          </div>
        </div>
        <Memo(AudioDevice)
          disabled={false}
          kind="audiooutput"
          onDeviceChange={[Function]}
          setDeviceError={[MockFunction]}
        />
        <Memo(AudioDevice)
          disabled={false}
          error=""
          kind="audioinput"
          onDeviceChange={[Function]}
          setDeviceError={[MockFunction]}
        />
        <div
          className="makeStyles-audioLevelContainer-3"
        >
          <div
            style={
              Object {
                "display": "flex",
                "justifyContent": "center",
                "width": "2em",
              }
            }
          >
            <Memo(Microphone) />
          </div>
          <ProgressBar
            duration={0.1}
            position={0}
            style={
              Object {
                "flex": "1",
              }
            }
          />
        </div>
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-confirmationButtons-7"
      item={true}
      lg={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        <strong>
           Does your audio sound good?
        </strong>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "1.5em",
          }
        }
        variant="contained"
      >
        Yes
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={false}
        onClick={[Function]}
      >
        Skip for now
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </div>
</WithStyles(ForwardRef(Container))>
`;
